crates/re_data_source/src/data_loader/mod.rs

```diff
@@ -0,0 +1,175 @@
+    ///
+    /// `path` isn't necessarily a _file_ path, but can be a directory as well: implementers are
+    /// free to handle that however they decide.
+    ///
+    /// ## Error handling
+    ///
+    /// Most implementers of `load_from_file` are expected to be asynchronous in nature.
+    ///
+    /// Asynchronous implementers should make sure to fail early (and thus synchronously) when
+    /// possible (e.g. didn't even manage to open the file).
+    /// Otherwise, they should log errors that happen in an asynchronous context.
+    ///
+    /// If a [`DataLoader`] has no interest in the given file, it should successfully return
+    /// without pushing any data into `tx`.
+    #[cfg(not(target_arch = "wasm32"))]
+    fn load_from_file(
+        &self,
+        store_id: re_log_types::StoreId,
+        path: std::path::PathBuf,
+        tx: std::sync::mpsc::Sender<LoadedData>,
+    ) -> Result<(), DataLoaderError>;
+
+    /// Loads data from in-memory file contents and sends it to `tx`.
+    ///
+    /// This is generally called when opening files via drag-and-drop or when using the web viewer.
+    ///
+    /// The passed-in `store_id` is a shared recording created by the file loading machinery:
+    /// implementers can decide to use it or not (e.g. it might make sense to log all images with a
+    /// similar name in a shared recording, while an rrd file is already its own recording).
+    ///
+    /// The `path` of the file is given for informational purposes (e.g. to extract the file's
+    /// extension): implementers should _not_ try to read from disk as there is likely isn't a
+    /// filesystem available to begin with.
+    /// `path` is guaranteed to be a file path.
+    ///
+    /// ## Error handling
+    ///
+    /// Most implementers of `load_from_file_contents` are expected to be asynchronous in nature.
+    ///
+    /// Asynchronous implementers should make sure to fail early (and thus synchronously) when
+    /// possible (e.g. didn't even manage to open the file).
+    /// Otherwise, they should log errors that happen in an asynchronous context.
+    ///
+    /// If a [`DataLoader`] has no interest in the given file, it should successfully return
+    /// without pushing any data into `tx`.
+    fn load_from_file_contents(
+        &self,
+        store_id: re_log_types::StoreId,
+        filepath: std::path::PathBuf,
+        contents: std::borrow::Cow<'_, [u8]>,
+        tx: std::sync::mpsc::Sender<LoadedData>,
+    ) -> Result<(), DataLoaderError>;
+}
+
+/// Errors that might happen when loading data through a [`DataLoader`].
+#[derive(thiserror::Error, Debug)]
+pub enum DataLoaderError {
+    #[cfg(not(target_arch = "wasm32"))]
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Arrow(#[from] re_log_types::DataCellError),
+
+    #[error(transparent)]
+    Decode(#[from] re_log_encoding::decoder::DecodeError),
+
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
```

Comment:
I can make `FileNotFound` a top-level error.

It won't affect `DataLoader` implementers, since the check for the file existence is done by the loading machinery before even dispatching to `DataLoader`s.
By the time we get to an actual `DataLoader`, `FileNotFound` should be incredibly rare (somebody removed the file between the check and the actual loader execution).

Reply:
Oh nevermind, I misread your comment.